File: static/js/tools/subnet-calc.js
(function () {
  const form = document.getElementById("tool-form");
  if (!form) return;
  const btn = document.getElementById("submit-btn");
  const result = document.getElementById("result");

  const FIELDS = [
    "network",
    "broadcast",
    "netmask",
    "wildcard",
    "first_host",
    "last_host",
    "usable_hosts",
    "total_hosts",
    "prefix",
    "ip_version",
  ];

  function renderResult(el, data) {
    const esc = WebTool.esc;
    const i18n = WebTool.i18n;
    let html = '<div class="result-card">';
    FIELDS.forEach((f) => {
      html +=
        "<p>" +
        esc(i18n("result.subnet." + f)) +
        ": <strong>" +
        esc(data[f]) +
        "</strong></p>";
    });
    html += "</div>";
    el.innerHTML = html;
  }

  form.onsubmit = async (e) => {
    e.preventDefault();
    WebTool.showLoading(btn);
    result.innerHTML = "";
    try {
      const res = await WebTool.api("subnet-calc", {
        cidr: document.getElementById("cidr").value.trim(),
      });
      renderResult(result, res.data);
      const dur = document.createElement("p");
      dur.className = "duration";
      dur.textContent = WebTool.fmtDuration(res.duration_ms);
      result.appendChild(dur);
    } catch (err) {
      WebTool.showError(result, err);
    } finally {
      WebTool.stopLoading(btn);
    }
  };
})();

File: services/subnet_calc.py
from __future__ import annotations


def _first_last_host(net) -> tuple[str, str]:
    if net.prefixlen == net.max_prefixlen:
        return str(net.network_address), str(net.network_address)
    if net.version == 4 and net.prefixlen == 31:
        return str(net.network_address), str(net.broadcast_address)
    return str(net.network_address + 1), str(net.broadcast_address - 1)


def _usable_hosts(net) -> int:
    if net.version == 4 and net.prefixlen <= 30:
        return net.num_addresses - 2
    return net.num_addresses


def calculate_subnet(net) -> dict:
    first_host, last_host = _first_last_host(net)
    return {
        "cidr": str(net),
        "network": str(net.network_address),
        "broadcast": str(net.broadcast_address),
        "netmask": str(net.netmask),
        "wildcard": str(net.hostmask),
        "first_host": first_host,
        "last_host": last_host,
        "usable_hosts": _usable_hosts(net),
        "total_hosts": net.num_addresses,
        "prefix": net.prefixlen,
        "ip_version": net.version,
    }
